p2p/raintree/router.go

```diff
@@ -282,6 +298,79 @@ func (rtr *rainTreeRouter) Size() int {
 	return rtr.peersManager.GetPeerstore().Size()
 }
 
+// handleStream ensures the peerstore contains the remote peer and then reads
+// the incoming stream in a new go routine.
+func (rtr *rainTreeRouter) handleStream(stream libp2pNetwork.Stream) {
+	rtr.logger.Debug().Msg("handling incoming stream")
+	peer, err := utils.PeerFromLibp2pStream(stream)
+	if err != nil {
+		rtr.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("parsing remote peer identity")
+
+		if err = stream.Reset(); err != nil {
+			rtr.logger.Error().Err(err).Msg("resetting stream")
+		}
+		return
+	}
+
+	if err := rtr.AddPeer(peer); err != nil {
+		rtr.logger.Error().Err(err).
+			Str("address", peer.GetAddress().String()).
+			Msg("adding remote peer to router")
+	}
+
+	go rtr.readStream(stream)
+}
+
+// readStream reads the incoming stream, extracts the serialized `PocketEnvelope`
+// data from the incoming `RainTreeMessage`, and passes it to the application by
+// calling the configured `rtr.handler`.
+func (rtr *rainTreeRouter) readStream(stream libp2pNetwork.Stream) {
+	// Time out if no data is sent to free resources.
+	if err := stream.SetReadDeadline(newReadStreamDeadline()); err != nil {
+		// NB: tests using libp2p's `mocknet` rely on this not returning an error.
+		// `SetReadDeadline` not supported by `mocknet` streams.
+		rtr.logger.Debug().Err(err).Msg("setting stream read deadline")
+	}
+
+	// log incoming stream
+	rtr.logStream(stream)
+
+	// read stream
+	data, err := io.ReadAll(stream)
+	if err != nil {
+		rtr.logger.Error().Err(err).Msg("reading from stream")
+		if err := stream.Reset(); err != nil {
+			rtr.logger.Debug().Err(err).Msg("resetting stream (read-side)")
+		}
+		return
+	}
+
+	// done reading; reset to signal this to remote peer
```

Comment:
Add a note that if we don't do this, we can easily max out the number of open network connections on the receiver's side